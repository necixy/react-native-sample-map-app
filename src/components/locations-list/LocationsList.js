import React, { Component } from "react";
import {
  View,
  Text,
  FlatList,
  TouchableOpacity,
  Image,
  Alert
} from "react-native";
import { PropTypes } from "prop-types";
import { graphql } from "react-apollo";
import locationsQuery from "../../graphql/queries/locations";
import deleteLocation from "../../graphql/mutations/deleteLocation";
import s from "./LocationsList.styles";

class LocationsList extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  editLocation(location) {
    this.props.navigation.navigate("EditLocation", { location });
  }

  promptDelete(location) {
    Alert.alert(
      "Confirm Delete",
      `Are you sure you would like to delete location: ${location.name}`,
      [
        { text: "Cancel", style: "cancel" },
        {
          text: "Delete",
          onPress: () => this.deleteLocation(location),
          style: "destructive"
        }
      ],
      { cancelable: false }
    );
  }

  async deleteLocation(location) {
    await this.props.mutate({
      variables: { id: location.id },
      refetchQueries: [{ query: locationsQuery }]
    });
  }

  renderErrorMsg(error) {
    return <Text style={s.errorMsg}>{error}</Text>;
  }

  renderListItem({ item, index }) {
    return (
      <View key={index} style={s.listItem}>
        <View style={s.itemDetails}>
          <Text>{item.name}</Text>
        </View>
        <View style={s.itemBtnContainer}>
          <TouchableOpacity
            style={s.itemBtn}
            onPress={this.editLocation.bind(this, item)}
          >
            <Image style={s.itemBtnIcon} source={require("./img/edit.png")} />
          </TouchableOpacity>
          <TouchableOpacity
            style={s.itemBtn}
            onPress={this.promptDelete.bind(this, item)}
          >
            <Image style={s.itemBtnIcon} source={require("./img/trash.png")} />
          </TouchableOpacity>
        </View>
      </View>
    );
  }

  renderList(locations) {
    const _keyExtractor = (item, index) => `location-${index}`;
    return (
      <FlatList
        {...this.props}
        style={[this.props.style, s.list]}
        data={locations}
        keyExtractor={_keyExtractor}
        renderItem={this.renderListItem.bind(this)}
      />
    );
  }

  render() {
    const { locations } = this.props;
    if (!locations || locations.length == 0) {
      return this.renderErrorMsg("No data available.");
    } else {
      return this.renderList(locations);
    }
  }
}

LocationsList.propTypes = {
  locations: PropTypes.array.isRequired,
  navigation: PropTypes.object.isRequired
};

export default graphql(deleteLocation)(LocationsList);
