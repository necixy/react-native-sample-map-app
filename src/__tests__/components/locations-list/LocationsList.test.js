import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { MockedProvider } from "react-apollo/test-utils";
import LocationsList from "../../../components/locations-list/LocationsList";

const locationsMocks = [
  {
    id: "5c40317ac907b9d99467ba08",
    name: "Brandenburg Tor",
    address: "Pariser Platz, 10117 Berlin, Germany",
    lat: 52.5162746,
    lng: 13.3777041
  },
  {
    id: "5c404e92c907b9d99467ba0b",
    name: "Mall of Berlin",
    address: "Leipziger Pl. 12, 10117 Berlin, Germany",
    lat: 52.51050919999999,
    lng: 13.3810452
  },
  {
    id: "5c404fd3c907b9d99467ba0c",
    name: "TV Tower",
    address: "Panoramastraße 1A, 10178 Berlin, Germany",
    lat: 52.520815,
    lng: 13.4094191
  },
  {
    id: "5c405010c907b9d99467ba0e",
    name: "Night club Tressor",
    address: "Köpenicker Str. 70, 10179 Berlin, Germany",
    lat: 52.5106031,
    lng: 13.4197619
  },
  {
    id: "5c405030c907b9d99467ba0f",
    name: "Technology museum Berlin",
    address: "Trebbiner Str. 9, 10963 Berlin, Germany",
    lat: 52.4986982,
    lng: 13.3778846
  },
  {
    id: "5c4054c4c907b9d99467ba11",
    name: "Alexanderplatz",
    address: "10178 Berlin, Germany",
    lat: 52.5219184,
    lng: 13.4132147
  }
];

describe("LocationsList Component Testing", () => {
  it("LocationsList Renders correctly", () => {
    const locationsList = renderer.create(
      <MockedProvider addTypename={false}>
        <LocationsList locations={locationsMocks} />
      </MockedProvider>
    );

    let listJSON = locationsList.toJSON();
    let root = locationsList.root;
    expect(listJSON).toMatchSnapshot();

    const list = root.find(
      el => el.props && el.props.testId == "list" && el.type == "RCTScrollView"
    );

    const itemNames = list.findAll(el => {
      return el.props && el.props.testId == "itemName" && el.type == "Text";
    });

    const itemEditBtns = list.findAll(el => {
      return el.props && el.props.testId == "itemEditBtn";
    });

    const itemDeleteBtns = list.findAll(el => {
      return el.props && el.props.testId == "itemDeleteBtn";
    });

    // Length of itemNames,itemEditBtns and itemDeleteBtns should be equal to length of locations.
    expect(itemNames.length == locationsMocks.length);
    expect(itemEditBtns.length == locationsMocks.length);
    expect(itemDeleteBtns.length == locationsMocks.length);

    // Each itemNames should show the location name.
    itemNames.forEach((el, i) => {
      el.children[0] == locationsMocks[i].name;
    });
  });
});
