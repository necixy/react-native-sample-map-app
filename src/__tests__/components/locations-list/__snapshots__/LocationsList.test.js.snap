// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationsList Component Testing LocationsList Renders correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "address": "Pariser Platz, 10117 Berlin, Germany",
        "id": "5c40317ac907b9d99467ba08",
        "lat": 52.5162746,
        "lng": 13.3777041,
        "name": "Brandenburg Tor",
      },
      Object {
        "address": "Leipziger Pl. 12, 10117 Berlin, Germany",
        "id": "5c404e92c907b9d99467ba0b",
        "lat": 52.51050919999999,
        "lng": 13.3810452,
        "name": "Mall of Berlin",
      },
      Object {
        "address": "Panoramastraße 1A, 10178 Berlin, Germany",
        "id": "5c404fd3c907b9d99467ba0c",
        "lat": 52.520815,
        "lng": 13.4094191,
        "name": "TV Tower",
      },
      Object {
        "address": "Köpenicker Str. 70, 10179 Berlin, Germany",
        "id": "5c405010c907b9d99467ba0e",
        "lat": 52.5106031,
        "lng": 13.4197619,
        "name": "Night club Tressor",
      },
      Object {
        "address": "Trebbiner Str. 9, 10963 Berlin, Germany",
        "id": "5c405030c907b9d99467ba0f",
        "lat": 52.4986982,
        "lng": 13.3778846,
        "name": "Technology museum Berlin",
      },
      Object {
        "address": "10178 Berlin, Germany",
        "id": "5c4054c4c907b9d99467ba11",
        "lat": 52.5219184,
        "lng": 13.4132147,
        "name": "Alexanderplatz",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  locations={
    Array [
      Object {
        "address": "Pariser Platz, 10117 Berlin, Germany",
        "id": "5c40317ac907b9d99467ba08",
        "lat": 52.5162746,
        "lng": 13.3777041,
        "name": "Brandenburg Tor",
      },
      Object {
        "address": "Leipziger Pl. 12, 10117 Berlin, Germany",
        "id": "5c404e92c907b9d99467ba0b",
        "lat": 52.51050919999999,
        "lng": 13.3810452,
        "name": "Mall of Berlin",
      },
      Object {
        "address": "Panoramastraße 1A, 10178 Berlin, Germany",
        "id": "5c404fd3c907b9d99467ba0c",
        "lat": 52.520815,
        "lng": 13.4094191,
        "name": "TV Tower",
      },
      Object {
        "address": "Köpenicker Str. 70, 10179 Berlin, Germany",
        "id": "5c405010c907b9d99467ba0e",
        "lat": 52.5106031,
        "lng": 13.4197619,
        "name": "Night club Tressor",
      },
      Object {
        "address": "Trebbiner Str. 9, 10963 Berlin, Germany",
        "id": "5c405030c907b9d99467ba0f",
        "lat": 52.4986982,
        "lng": 13.3778846,
        "name": "Technology museum Berlin",
      },
      Object {
        "address": "10178 Berlin, Germany",
        "id": "5c4054c4c907b9d99467ba11",
        "lat": 52.5219184,
        "lng": 13.4132147,
        "name": "Alexanderplatz",
      },
    ]
  }
  maxToRenderPerBatch={10}
  mutate={[Function]}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      undefined,
      Object {
        "borderTopColor": "#ccc",
        "borderTopWidth": 1,
        "marginTop": 10,
      },
    ]
  }
  testId="list"
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#ccc",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginTop": 2,
            "minHeight": 40,
            "paddingHorizontal": 20,
            "paddingVertical": 15,
          }
        }
        testClass="listItem"
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            testId="itemName"
          >
            Brandenburg Tor
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/edit.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/trash.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#ccc",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginTop": 2,
            "minHeight": 40,
            "paddingHorizontal": 20,
            "paddingVertical": 15,
          }
        }
        testClass="listItem"
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            testId="itemName"
          >
            Mall of Berlin
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/edit.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/trash.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#ccc",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginTop": 2,
            "minHeight": 40,
            "paddingHorizontal": 20,
            "paddingVertical": 15,
          }
        }
        testClass="listItem"
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            testId="itemName"
          >
            TV Tower
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/edit.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/trash.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#ccc",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginTop": 2,
            "minHeight": 40,
            "paddingHorizontal": 20,
            "paddingVertical": 15,
          }
        }
        testClass="listItem"
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            testId="itemName"
          >
            Night club Tressor
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/edit.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/trash.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#ccc",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginTop": 2,
            "minHeight": 40,
            "paddingHorizontal": 20,
            "paddingVertical": 15,
          }
        }
        testClass="listItem"
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            testId="itemName"
          >
            Technology museum Berlin
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/edit.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/trash.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#ccc",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginTop": 2,
            "minHeight": 40,
            "paddingHorizontal": 20,
            "paddingVertical": 15,
          }
        }
        testClass="listItem"
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            testId="itemName"
          >
            Alexanderplatz
          </Text>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/edit.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 10,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/components/locations-list/img/trash.png",
                }
              }
              style={
                Object {
                  "height": 25,
                  "marginHorizontal": 5,
                  "width": 25,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
